### PR TITLE
Explicitly define Sidecar Resources

### PR DESCRIPTION
**Pull request by chuckydev**
Required by #243 

**Currently**

Sidecar resources are currently defined by the pod size and floored to 50Mi and bursting CPU. We would like all resources to be more explicit and part of the cassandra resource itself

**End Approach**

Sidecar `ìmage` and `resources` have now been added to their own block in the sky-uk cassandra resource as a Child of `pod`. The same validation as other resources has been included.

```
  pod:
    bootstrapperImage: 
    image:
    readinessProbe:
      timeoutSeconds: 
    resources:
      limits:
        memory: 
      requests:
        cpu: 
        memory: 
    sidecar:
      image: 
      resources:
        requests:
          cpu: 
          memory: 
        limits:
          memory: 
```

This is a **breaking change** as sidecarImage has been moved to `sidecar.image`, resources are mandatory. `sidecar.image` remains defaulted and optional.

A change will also be made to the wiki to describe this change.

I have also added a fix to genereate-all which was currently not working due to a bad path. 
https://github.com/kubernetes/sample-apiserver#when-using-go-111-modules
https://github.com/kubernetes/code-generator/issues/57

### CODE REVIEW

File: cassandra-operator/hack/update-codegen.sh [hunk 1]
```diff
@@ -15,9 +15,18 @@ groupVersion="cassandra:v1alpha1"
...
-${codeGenDir}/generate-groups.sh all \
+# Ensure that cassandra-operator is cloned under you GOPATH like so
+# $GOPATH/github.com/sky-uk/cassandra-operator
+# This is to ensure code-generator can be run from inside the vendor directory.
+${codeGenDir}/generate-groups.sh deepcopy \
   ${basePackage}/pkg/client ${basePackage}/pkg/apis \
   ${groupVersion} \
-  --go-header-file ${scriptDir}/empty-boilerplate.txt
+  --go-header-file ${scriptDir}/empty-boilerplate.txt \
+  --output-base ${projectDir}/../../../..
```
Comment (Bez625) [inline cassandra-operator/hack/update-codegen.sh:32]: Is it possible to replace the relative path with something more explicit?

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/types_test.go [hunk 2]
```diff
@@ -36,8 +36,8 @@ var _ = Describe("Cassandra Types", func() {
 			podInequalityCheck,
 			Entry("when one pod has a nil bootstrap image", func(pod, _ *Pod) { pod.BootstrapperImage = nil }),
 			Entry("when pods have different bootstrap images", func(pod, _ *Pod) { pod.BootstrapperImage = ptr.String("another image") }),
-			Entry("when one pod has a nil sidecar image", func(pod, _ *Pod) { pod.SidecarImage = nil }),
-			Entry("when pods have different sidecar images", func(pod, _ *Pod) { pod.SidecarImage = ptr.String("another image") }),
+			Entry("when one pod has a nil sidecar image", func(pod, _ *Pod) { pod.Sidecar.Image = nil }),
+			Entry("when pods have different sidecar images", func(pod, _ *Pod) { pod.Sidecar.Image = ptr.String("another image") }),
```
Comment (Bez625) [inline cassandra-operator/pkg/apis/cassandra/v1alpha1/types_test.go:40]: Are these needed? It looks like they are repeated in the sidecar specific checks below

File: cassandra-operator/pkg/cluster/cluster.go [hunk 3]
```diff
@@ -478,21 +473,7 @@ func (c *Cluster) createCassandraSidecarContainer(rack *v1alpha1.Rack) v1.Contai
...
-				v1.ResourceMemory: minQuantity(
-					*c.definition.Spec.Pod.Resources.Requests.Memory(),
-					maxSidecarMemoryRequest,
-				),
-			},
-			Limits: v1.ResourceList{
-				v1.ResourceMemory: sidecarMemoryLimit,
-			},
-		},
+		Resources: *&c.definition.Spec.Pod.Sidecar.Resources,
```
Comment (Bez625) [inline cassandra-operator/pkg/cluster/cluster.go:476]: Should this not be `c.definition.Spec.Pod.Sidecar.Resources` ?

File: cassandra-operator/test/apis/builder.go [hunk 4]
```diff
@@ -428,6 +432,53 @@ func (p *PodSpecBuilder) Build() *v1alpha1.Pod {
 	}
 }
 
+//
+// SidecarBuilder
+//
+
+type SidecarBuilder interface {
```
Comment (Bez625) [inline cassandra-operator/test/apis/builder.go:439]: Is this used anywhere?

File: cassandra-snapshot/test/e2e/steps.go [hunk 5]
```diff
@@ -246,7 +246,7 @@ func SnapshotListForPod(pod *v1.Pod) ([]Snapshot, error) {
 
 func BackdateSnapshotsForPods(pods []*v1.Pod, backdatePeriod time.Duration) {
 	newSnapshotName := time.Now().Unix() - int64(backdatePeriod.Seconds())
-	renameSnapshot(pods, string(newSnapshotName))
+	renameSnapshot(pods, fmt.Sprintf("%d", newSnapshotName))
```
Comment (Bez625) [inline cassandra-snapshot/test/e2e/steps.go:249]: Is this to remove the casting? I'm assuming it is just code quality and the format is the same?

File: cassandra-operator/test/e2e/parallel/modification/sidecar_external_modifiation_test.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		if CurrentGinkgoTestDescription().Failed {
+			PrintDiagnosis(Namespace, testStartTime, clusterName)
+		}
+	})
+
+	AfterEach(func() {
+		DeleteCassandraResourcesForClusters(Namespace, clusterName)
+	})
+
+	It("should eventually return the sidecar container definitions back desired state if the sidecar image is changed", func() {
```
Comment (Bez625) [inline cassandra-operator/test/e2e/parallel/modification/sidecar_external_modifiation_test.go:48]: Is the word "back" in this incorrectly?

**Review by Bez625 — CHANGES_REQUESTED** [review]

File: cassandra-operator/hack/update-codegen.sh (hunk 1 above)
Comment (chuckydev) [reply cassandra-operator/hack/update-codegen.sh:32]: Unfortunately not. I know its not great, but there are real limitations with the path being in a `github.com/sky-uk/cassandra-operator` folder. And going from the script dir I think thats the best route. Its also how kubernetes do this in sample-apiserver which is a reference app for code-generator

https://github.com/kubernetes/sample-apiserver#when-using-go-111-modules
kubernetes/code-generator#57

File: cassandra-operator/pkg/apis/cassandra/v1alpha1/types_test.go (hunk 2 above)
Comment (chuckydev) [reply cassandra-operator/pkg/apis/cassandra/v1alpha1/types_test.go:40]: I left them in as they still have a function of prooving that sidecar is involved in Pod equality.

File: cassandra-operator/test/apis/builder.go [hunk 7]
```diff
@@ -428,6 +432,49 @@ func (p *PodSpecBuilder) Build() *v1alpha1.Pod {
...
+
+func (s *SidecarSpecBuilder) WithDefaults() *SidecarSpecBuilder {
+	s.image = ptr.String("SidecarImage")
+	s.resources = &coreV1.ResourceRequirements{
+		Limits: coreV1.ResourceList{
+			coreV1.ResourceMemory: resource.MustParse("50Mi"),
+		},
+		Requests: coreV1.ResourceList{
+			coreV1.ResourceMemory: resource.MustParse("50Mi"),
+			coreV1.ResourceCPU:    resource.MustParse("100m"),
```
Comment (sebbonnet) [inline cassandra-operator/test/apis/builder.go:451]: we should probably default to bursting (e.g. 0 request)

**Review by sebbonnet — CHANGES_REQUESTED** [review]
lgtm - just a few minor comments to address

File: cassandra-operator/test/e2e/parallel/modification/modification_test.go [hunk 8]
```diff
@@ -87,8 +87,19 @@ var _ = Context("Allowable cluster modifications", func() {
...
 		TheClusterPodSpecAreChangedTo(Namespace, clusterName, v1alpha1.Pod{
 			BootstrapperImage: CassandraBootstrapperImageName,
-			SidecarImage:      CassandraSidecarImageName,
-			Image:             CassandraImageName,
+			Sidecar: &v1alpha1.Sidecar{
+				Image: CassandraSidecarImageName,
+				Resources: coreV1.ResourceRequirements{
+					Requests: coreV1.ResourceList{
+						coreV1.ResourceMemory: resource.MustParse("50Mi"),
+						coreV1.ResourceCPU:    resource.MustParse("100m"),
```
[the remark this answers is not in the document]
Comment (chuckydev) [reply cassandra-operator/test/e2e/parallel/modification/modification_test.go]: Ive now asserted the sidecar resources as part of a modification test just to verify.

File: cassandra-operator/test/e2e/parallel/creation/creation_test.go [hunk 9]
```diff
@@ -82,8 +82,19 @@ func createClustersInParallel(multipleRacksCluster, emptyDirCluster *TestCluster
...
-				Image:             CassandraImageName,
+				Sidecar: &v1alpha1.Sidecar{
+					Image: CassandraSidecarImageName,
+					Resources: coreV1.ResourceRequirements{
+						Requests: coreV1.ResourceList{
+							coreV1.ResourceMemory: resource.MustParse("50Mi"),
+							coreV1.ResourceCPU:    resource.MustParse("100m"),
+						},
+						Limits: coreV1.ResourceList{
+							coreV1.ResourceMemory: resource.MustParse("50Mi"),
```
[the remark this answers is not in the document]
Comment (chuckydev) [reply cassandra-operator/test/e2e/parallel/creation/creation_test.go:93]: Ive now asserted the sidecar resources as part of the happy case creation test

File: cassandra-operator/hack/update-codegen.sh [hunk 10]
```diff
@@ -25,7 +25,7 @@ chmod +x ${codeGenDir}/generate-groups.sh
 # Ensure that cassandra-operator is cloned under you GOPATH like so
 # $GOPATH/github.com/sky-uk/cassandra-operator
 # This is to ensure code-generator can be run from inside the vendor directory.
-${codeGenDir}/generate-groups.sh deepcopy \
+${codeGenDir}/generate-groups.sh all \
```
Comment (sebbonnet) [inline cassandra-operator/hack/update-codegen.sh:28]: good catch! :)

**Review by sebbonnet — APPROVED** [review]

**Review by Bez625 — APPROVED** [review]

File: cassandra-operator/hack/update-codegen.sh (hunk 10 above)
Comment (chuckydev) [reply cassandra-operator/hack/update-codegen.sh:28]: Have to put this down to a good eye by @Bez625 :)

**Pull request merged**